llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp

```diff
@@ -2326,6 +2326,18 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
     }
 #endif
 
+    if (ST->isPreciseMemoryEnabled() && Inst.mayLoadOrStore()) {
+      ++Iter;
+      if (!isWaitInstr(*Iter)) {
```

Comment:
This isn't right. This will prevent adding youtr waitcnt if *any* waitcnt follows, even a wait for a different counter.

Normally `generateWaitcnt` should combine and remove redundant waitcnts like the ones in your test. It looks like the problem is that this does not normally happen at the end of a basic block. I have tried fixing that in #87539. If that patch is committed, you should be able to rebase this patch and the duplicate waitcnts will get removed.

Reply:
#87539 has been merged. You can merge-or-rebase main into your branch and remove this `isWaitInstr` check.